python/ql/src/semmle/python/Pruning.qll

```diff
@@ -0,0 +1,605 @@
+        }
+
+        override predicate impliesFalse(Constraint other) {
+            other.constrainsVariableToBe(this.booleanValue().booleanNot())
+        }
+
+    }
+
+    /** A constraint that the variable is None `(var is None) is True` or not None `(var is None) is False` */
+    class IsNone extends Constraint, TIsNone {
+
+        private boolean isNone() {
+            this = TIsNone(result)
+        }
+
+        override string toString() {
+            result = "Is None" and this.isNone() = true
+            or
+            result = "Is not None" and this.isNone() = false
+        }
+
+        override Constraint invert() {
+            result = TIsNone(this.isNone().booleanNot())
+        }
+
+        override predicate constrainsVariableToBe(boolean value) {
+            value = false and this.isNone() = true
+        }
+
+        override predicate impliesFalse(Constraint other) {
+            other = TIsNone(this.isNone().booleanNot())
+            or
+            this.isNone() = true and other = TTruthy(true)
+        }
+
+    }
+
+    /** A constraint that the variable fulfils some equality or inequality to an integral constant.
+     * `(var op k) is True` where `op` is an equality or inequality operator and `k` is an integer constant
+     */
+    class ConstrainedByConstant extends Constraint, TConstrainedByConstant {
+
+        private int intValue() {
+            this = TConstrainedByConstant(_, result)
+        }
+
+        private CompareOp getOp() {
+            this = TConstrainedByConstant(result, _)
+        }
+
+        override string toString() {
+            result = this.getOp().repr() + " " + this.intValue().toString()
+        }
+
+        override Constraint invert() {
+            result = TConstrainedByConstant(this.getOp().invert(), this.intValue())
+        }
+
+        override predicate constrainsVariableToBe(boolean value) {
+            this.getOp() = eq() and this.intValue() = 0 and value = false
+            or
+            value = true and (
+                this.getOp() = eq() and this.intValue() != 0
+                or
+                this.getOp() = ne() and this.intValue() = 0
+                or
+                this.getOp() = lt() and this.intValue() <= 0
+                or
+                this.getOp() = le() and this.intValue() < 0
+                or
+                this.getOp() = gt() and this.intValue() >= 0
+                or
+                this.getOp() = ge() and this.intValue() > 0
+            )
+        }
+
+        override predicate impliesFalse(Constraint other) {
+            exists(boolean b |
+                this.constrainsVariableToBe(b) and other = TTruthy(b.booleanNot())
+            )
+            or
+            this.getOp() = eq() and other = TIsNone(true)
+            or
+            this.getOp() = ne() and other.(ConstrainedByConstant).getOp() = eq()
+            and this.intValue() = other.(ConstrainedByConstant).intValue()
```

Comment:
Performance, the set of all bounds that contradict each other is rather large. So we want to check inequalities only when relevant.

Reply:
In that case, would it not be even better to include the constrained variable in `TConstrainedByConstant`? That way, the `contradicts` could check that it's in fact the same variable that's being constrained.

Reply:
I don't think so. Why have a " `x` is true" for every `x` when we just need one "is true"?

Reply:
Only the `ConstrainedByConstant` constraints would need the variable.

Consider this. If I have
```
x1 <= 1
x2 <= 2
...
x10 <= 10

y1 >= 101
y2 >= 102
...
y10 >= 110
```
in my source code, I will get a `TConstrainedByConstant` for each of the above. Now, each `x` will contradict every `y`, so `contradicts` on line 552 may contain all 100 tuples corresponding to each combination of `x` and `y` (depending on how well the magic is applied).

Perhaps it would make sense to add an `exists` that checks that the `this` and `other` constraints are both associated with the same variable?

Finally, as written, the above predicate is not symmetric. Perhaps add `or other.contradicts(this)` at the end?

Reply:
It isn't symmetric. 
`x is None` ⇒ `x is Falsey` but `x is Falsey` ⇏ `x is None`
Perhaps `impliesFalse` was a better name.

Reply:
I don't understand your argument. What is `x`? Is it `this` or `other` (or both)?
Judging by the documentation for `contradicts`/`impliesFalse`, I would certainly expect it to be symmetric. If A(x) being true implies B(x) is false, that means "A(x) and B(x)" is false. But then "B(x) and A(x)" is similarly false, and moving backwards B(x) true implies A(x) false.

Reply:
`x` is a Python variable, as in:
```python
if x:
   if x is None:
       #unreachable
if not x:
     if x is not None:
          #reachable
```
I think I misread you as saying than if  `A ⇒ not B` then `not A ⇒ B`, rather than if `A ⇒ not B` then `B ⇒ not A`.

All inequalities are handled below on line 552, so I don't think we want the additional disjunct here.
I'll add a `pragma [inline]` below to make the intention clear (the compiler inlines it already).

Reply:
I'm not sure a comment for the predicate on line 552 is sufficient. My whole point was that the above predicate, as currently written, looks incomplete ("where's the bounds checking?") and that a forward comment pointing to the place where this takes place would help.

Also, the above predicate still isn't symmetric: If `this = TIsNone(true)` and `other` is `eq` `0` (the second disjunct with `this` and `other` swapped, for the specific value `0`), then I don't believe it'll work, as the test in `contradicts` for `IsNone` checks for `other = TTruthy(true)`. To fix this, I think it should be sufficient to add `or other instanceof ConstrainedByConstant` after the `TTruthy` test.

Reply:
My bad, the `instanceof` check isn't sufficient, as it has to be `eq` specifically that's the operator.